### PR TITLE
[heft] Add the ability for plugins to specify custom actions.

### PR DESCRIPTION
**Pull request by iclanton**

### CODE REVIEW

File: apps/heft/src/cli/actions/CustomAction.ts [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    );
+
+    this._customActionOptions = customActionOptions;
+  }
+
+  public onDefineParameters(): void {
+    super.onDefineParameters();
+
+    this._parameterValues = new Map<string, () => CustomActionParameterType>();
+    for (const paramterOption of this._customActionOptions.parameters || []) {
```
Comment (octogonz) [inline apps/heft/src/cli/actions/CustomAction.ts:57]: `paramterOption` is mispelled

File: apps/heft/src/cli/actions/CustomAction.ts [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    for (const parameterOption of this._customActionOptions.parameters || []) {
+      if (this._parameterValues.has(parameterOption.callbackValueName)) {
+        throw new Error(`Duplicate callbackValueName: ${parameterOption.callbackValueName}`);
+      }
+
+      let getParameterValue: () => CustomActionParameterType;
+
+      switch (parameterOption.kind) {
+        case 'flag': {
+          const parameter: CommandLineFlagParameter = this.defineFlagParameter({
```
Comment (octogonz) [inline apps/heft/src/cli/actions/CustomAction.ts:66]: Here's a couple other scenarios to think about:

- My plugin implements localization as a preprocessor that runs before the TypeScript compiler task.  Can my plugin add a `--locale LOCALE` parameter to every existing Heft action that invokes the compiler?
- My plugin adds a new action `heft deploy` that performs build+test as part of deploying. Can it reuse some existing parameter definitions (e.g. `--production`, `--lite`, `--max-write-parallelism`) without having to copy+paste the code that defines those parameters?

Somehow this design problem feels very similar to the config file problem... like they should have similar solutions.

But I'm fine with getting a simple solution out while we think through the best design.

File: build-tests/heft-action-plugin/src/index.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+      documentation: 'An example custom action',
+      parameters: [
+        {
+          kind: 'flag',
+          paramterLongName: '--production',
+          description: 'Run in production mode',
+          callbackValueName: 'production'
+        }
+      ],
+      callback: ({ production }) => {
```
Comment (octogonz) [inline build-tests/heft-action-plugin/src/index.ts:21]: 🤔 We need type safety here.

File: apps/heft/src/cli/actions/CustomAction.ts (hunk 2 above)
Comment (iclanton) [reply apps/heft/src/cli/actions/CustomAction.ts:104]: I think that's a good north star, and I think that'll involve overhauling the whole action system in Heft to do right.

I'd like to get this implementation out and start playing with it, and once we've gained some knowledge and experience with how this all should work, we revisit things.

BTW - I marked these APIs as `@alpha`. Do you know if api-extractor trims `@alpha`s?

File: apps/heft/src/cli/actions/CustomAction.ts [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+export interface ICustomActionParameterString extends ICustomActionParameterBase<string> {
+  kind: 'string';
+}
+
+/** @alpha */
+export interface ICustomActionParameterStringList extends ICustomActionParameterBase<ReadonlyArray<string>> {
+  kind: 'stringList';
+}
+
+/** @alpha */
```
Comment (octogonz) [inline apps/heft/src/cli/actions/CustomAction.ts:33]: The Rush Stack repo doesn't have a support model that would meaningfully distinguish `@alpha` versus `@beta`.  Should probably do a PR to bulk search+replace `@alpha` with `@beta` everywhere.

We should probably also search for any APIs that have stagnated as `@beta` for more than a year and consider promoting them to `@public`.

File: build-tests/heft-action-plugin-test/README.md [hunk 5]
```diff
@@ -0,0 +1,6 @@
+# heft-action-plugin-test
+
+This project folder is one of the **build-tests** for the [Heft](https://www.npmjs.com/package/@rushstack/heft)
+build system. This project exercises a custom Heft action.
+
+Please see the [Heft documentation](http://localhost:4000/pages/heft/overview/) for documentation and tutorials.
```
Comment (octogonz) [inline build-tests/heft-action-plugin-test/README.md:6]: This URL is incorrect. If it was copy+pasted from somewhere, we should fix it everywhere.

**Review by octogonz — APPROVED** [review]

**Pull request merged**